lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/StateMachine/DriverTaskState.cs

```diff
@@ -0,0 +1,136 @@
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.FailedTaskCommunicationError), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskFailedBySystemError, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskFailedByGroupCommunication, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure }
+        });
+
+        private readonly static ISet<TaskState> FinalState = new HashSet<TaskState>()
+        {
+            TaskState.TaskFailedByAppError,
+            TaskState.TaskFailedBySystemError,
+            TaskState.TaskFailedByEvaluatorFailure,
+            TaskState.TaskFailedByGroupCommunication,
+            TaskState.TaskClosedByDriver,
+            TaskState.TaskCompleted
+        };
+
+        private volatile TaskState _currentState;
+
+        /// <summary>
+        /// Create a new DriverTaskState with TaskNew as the task initial state
+        /// </summary>
+        internal DriverTaskState()
+        {
+            _currentState = TaskState.TaskNew;
+        }
+
+        /// <summary>
+        /// return the current task state
+        /// </summary>
+        internal TaskState CurrentState
+        {
+            get
+            {
+                return _currentState;
+            }
+        }
+
+        /// <summary>
+        /// Get next valid state based on the current state and event given without changing the current state
+        /// If there is no valid next state, TaskStateTransitionException will be thrown.
+        /// </summary>
+        /// <param name="taskEvent"></param>
+        /// <returns></returns>
+        internal TaskState GetNext(TaskStateEvent taskEvent)
+        {
+            StateTransition<TaskState, TaskStateEvent> transition = new StateTransition<TaskState, TaskStateEvent>(_currentState, taskEvent);
+            TaskState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
+            {
+                throw new TaskStateTransitionException(_currentState, taskEvent);
+            }
+            return nextState;
+        }
+
+        /// <summary>
+        /// Move to the next state
+        /// If it is not able to move to the next valid state for a given event, TaskStateTransitionException will be thrown.
+        /// </summary>
+        /// <param name="taskEvent"></param>
+        /// <returns></returns>
+        internal TaskState MoveNext(TaskStateEvent taskEvent)
+        {
+            _currentState = GetNext(taskEvent);
+            return _currentState;
+        }
+
+        /// <summary>
+        /// Checks if the current state is a final state
+        /// </summary>
+        /// <returns></returns>
+        internal bool IsFinalState()
+        {
+            return FinalState.Contains(_currentState);
```

Comment:
Please replace with `IsFinalState(_currentState)`, so that all logic of deciding whether a state is final is in one place.